--- main.c.org	2025-07-11 00:43:36.964606686 +0900
+++ main.c	2025-07-11 00:45:32.232191578 +0900
@@ -1,4 +1,5 @@
 #include <stdio.h>
+#include <stdlib.h>
 #include "revappi.h"
 #include "romsrc.c"
 static const thunk_type thunk_int =
@@ -148,6 +149,12 @@
     PRIMITIVE(getc),
     PRIMITIVE_END
 };
+static cell *runout(void)
+{
+    static size_t count = 0;
+    fprintf(stderr, "%u\n", ++count);
+    return malloc(sizeof(cell));
+}
 #define RAM_SIZE 100000
 static cell ram[RAM_SIZE];
 static int place_source(const char *path)
@@ -187,8 +194,7 @@
     fclose(file);
     if (0 < brac)
 	return 2;
-    cell_allocator_init((cell *) ((((size_t) sip) + sizeof(cell) - 1) &
-				  -sizeof(cell)), &ram[RAM_SIZE], NULL);
+    cell_allocator_init(NULL, NULL, runout);
     return 0;
 }
 int main(int argc, char **argv)
