--- revappi.h	2023-09-27 00:54:48.247122109 +0900
+++ revappi.h.new	2023-09-27 00:37:28.167736542 +0900
@@ -1,4 +1,5 @@
 #include <stddef.h>
+#include "system.h"
 #define PRIM_REGFILE_SIZE 10
 typedef struct thunk thunk;
 typedef union {
--- revappi.c	2023-09-27 00:54:48.257121901 +0900
+++ revappi.c.new	2023-09-27 00:48:22.755034392 +0900
@@ -1,11 +1,8 @@
-#include <assert.h>
-#include <stdio.h>
 #include "revappi.h"
 #define RAM_SIZE 100000
 #define PRIM_REGFILE_SIZE 10
 static cell ram[RAM_SIZE];
-static cell *free_cell = NULL, *cell_for_chunk, *cell_used =
-    &ram[RAM_SIZE];
+static cell *free_cell = NULL, *cell_for_chunk, *cell_used;
 cell *cell_alloc(void)
 {
     cell *rtn = free_cell;
@@ -13,9 +10,6 @@
 	free_cell = rtn->free;
     else if (cell_for_chunk < cell_used)
 	rtn = --cell_used;
-#ifdef ALLOC_TRACE
-    fprintf(stderr, "@ a.out + %p 0x%x\n", (void *) rtn, sizeof(cell));
-#endif
     return rtn;
 }
 void cell_free(void *cellp)
@@ -23,9 +17,6 @@
     cell *cp = cellp;
     cp->free = free_cell;
     free_cell = cp;
-#ifdef ALLOC_TRACE
-    fprintf(stderr, "@ a.out - %p\n", (void *) cp);
-#endif
 }
 static thunk *thunk_alloc(void)
 {
@@ -373,12 +364,12 @@
 static int place_source(const char *path, char **sip_p)
 {
     cell *cur = ram;
-    FILE *file = fopen(path, "r");
+    int file = my_open(path);
     char *sip = ram[0].m;
     size_t brac = 0;
     while (1) {
 	char *curp = cur->m;
-	size_t i, byte = fread(curp, 1, sizeof(cell), file);
+	size_t i, byte = my_read(file, curp, sizeof(cell));
 	for (i = 0; i < byte; i++)
 	    switch (curp[i]) {
 	    case '(':
@@ -396,7 +387,7 @@
 	}
 	cur++;
     }
-    fclose(file);
+    my_close(file);
     if (0 < brac)
 	return 2;
     cell_for_chunk = cur + 1;
@@ -433,6 +424,12 @@
     }
     return rtn;
 }
+static void print_error(const char *msg)
+{
+    size_t msg_len;
+    for (msg_len = 0; '\0' != msg[msg_len]; msg_len++);
+    my_write(my_stderr, msg, msg_len);
+}
 int main_core(const prim_env_member *membs, const char *romsrc, int argc,
 	      char **argv)
 {
@@ -442,9 +439,11 @@
     force_regfile r;
     list *stack = NULL;
     if (2 != argc) {
-	fprintf(stderr, "%s [source file]\n", argv[0]);
+	print_error(argv[0]);
+	print_error(" [source file]\n");
 	return 1;
     }
+    cell_used = &ram[RAM_SIZE];
     rst = place_source(argv[1], &src);
     if (rst)
 	return rst + 1;
@@ -452,7 +451,6 @@
     top.c.src.env = gen_prim_env(membs);
     top.c.src.sip = romsrc;
     rst = exec_src(&stack, &top.c.src.env, &top.c.src.sip);
-    assert(exec_src_nop == rst && NULL == stack);
     top.c.src.sip = src;
     thp = force(&r, &top);
     return !(&top == thp && NULL == top.c.force.thp
--- main.c	2023-09-27 00:54:48.287121279 +0900
+++ main.c.new	2023-09-27 00:50:05.162995704 +0900
@@ -1,4 +1,3 @@
-#include <stdio.h>
 #include "revappi.h"
 #include "romsrc.c"
 static const thunk_type thunk_int =
@@ -52,6 +51,7 @@
     { PIP_POP_THUNK, PIP_CALL(name), PIP_DIST(int) }
 CONSTINT(0, zero);
 CONSTINT(1, one);
+#define EOF (-1)
 CONSTINT(EOF, eof);
 #define COMPAREINT(OP,name) \
 static size_t beta_prim_##name(word *r)\
@@ -93,9 +93,10 @@
     PIP_POP_THUNK, PIP_POP(world), PIP_POP(world), PIP_CALL(joinworld),
     PIP_DIST(world)
 };
-static size_t beta_primcore_putc(FILE *f, word *r)
+static size_t beta_primcore_putc(int fd, word *r)
 {
-    fputc(r[1].i, f);
+    char c = r[1].i;
+    my_write(fd, &c, 1);
     r[1].thp = r[2].thp;
     return 0;
 }
@@ -108,12 +109,16 @@
     PIP_POP_THUNK, PIP_POP(int), PIP_POP(world),\
     PIP_CALL(name), PIP_DIST(world)\
 }
-PUTC_CORE(stdout, putc);
-PUTC_CORE(stderr, errc);
+PUTC_CORE(my_stdout, putc);
+PUTC_CORE(my_stderr, errc);
 static size_t beta_prim_getc(word *r)
 {
+    unsigned char c;
     r[2].thp = r[1].thp;
-    r[1].i = fgetc(stdin);
+    if (1 != my_read(my_stdin, &c, 1))
+	r[1].i = EOF;
+    else
+	r[1].i = c;
     return 0;
 }
 static const thunk_type *const dist_getc[] =
@@ -126,9 +131,7 @@
     PRIMITIVE(equal), PRIMITIVE(big), PRIMITIVE(eqbig),
     PRIMITIVE(startworld), PRIMITIVE(forkworld), PRIMITIVE(joinworld),
     PRIMITIVE(putc),
-#ifndef ALLOC_TRACE
     PRIMITIVE(errc),
-#endif
     PRIMITIVE(getc),
     PRIMITIVE_END
 };
