--- revappi.c	2023-11-18 21:54:43.038659415 +0900
+++ revappi.c.new	2023-11-18 21:44:58.640115029 +0900
@@ -1,7 +1,3 @@
-#include <assert.h>
-#ifdef ALLOC_TRACE
-#include <stdio.h>
-#endif
 #include "revappi.h"
 static cell *free_cell =
     NULL, *cell_spare, *cell_limit, *(*cell_runout)(void);
@@ -20,9 +16,6 @@
 	rtn = cell_spare++;
     else
 	rtn = cell_runout();
-#ifdef ALLOC_TRACE
-    fprintf(stderr, "@ a.out + %p 0x%x\n", (void *) rtn, sizeof(cell));
-#endif
     return rtn;
 }
 void cell_free(void *cellp)
@@ -30,9 +23,6 @@
     cell *cp = cellp;
     cp->free = free_cell;
     free_cell = cp;
-#ifdef ALLOC_TRACE
-    fprintf(stderr, "@ a.out - %p\n", (void *) cp);
-#endif
 }
 static thunk *thunk_alloc(void)
 {
@@ -412,7 +402,6 @@
     top.c.src.env = gen_prim_env(membs);
     top.c.src.sip = romsrc;
     rst = exec_src(&stack, &top.c.src.env, &top.c.src.sip);
-    assert(exec_src_nop == rst && NULL == stack);
     top.c.src.sip = ramsrc;
     thp = force(&r, &top);
     return !(&top == thp && NULL == top.c.force.thp
--- main.c	2023-11-18 21:54:43.058659027 +0900
+++ main.c.new	2023-11-18 21:54:17.029163725 +0900
@@ -1,4 +1,4 @@
-#include <stdio.h>
+#include "system.h"
 #include "revappi.h"
 #include "romsrc.c"
 static const thunk_type thunk_int =
@@ -68,6 +68,7 @@
     { PIP_POP_THUNK, PIP_CALL(name), PIP_DIST(int) }
 CONSTINT(0, zero);
 CONSTINT(1, one);
+#define EOF (-1)
 CONSTINT(EOF, eof);
 #define COMPAREINT(OP,name) \
 static size_t beta_prim_##name(word *r)\
@@ -109,9 +110,10 @@
     PIP_POP_THUNK, PIP_POP(world), PIP_POP(world), PIP_CALL(joinworld),
     PIP_DIST(world)
 };
-static size_t beta_primcore_putc(FILE *f, word *r)
+static size_t beta_primcore_putc(int fd, word *r)
 {
-    fputc(word_to_int(r[1]), f);
+    char c = word_to_int(r[1]);
+    my_write(fd, &c, 1);
     r[1].thp = r[2].thp;
     return 0;
 }
@@ -124,12 +126,18 @@
     PIP_POP_THUNK, PIP_POP(int), PIP_POP(world),\
     PIP_CALL(name), PIP_DIST(world)\
 }
-PUTC_CORE(stdout, putc);
-PUTC_CORE(stderr, errc);
+PUTC_CORE(my_stdout, putc);
+PUTC_CORE(my_stderr, errc);
 static size_t beta_prim_getc(word *r)
 {
+    unsigned char c;
+    int i;
     r[2].thp = r[1].thp;
-    r[1] = int_to_word(fgetc(stdin));
+    if (1 != my_read(my_stdin, &c, 1))
+	i = EOF;
+    else
+	i = c;
+    r[1] = int_to_word(i);
     return 0;
 }
 static const thunk_type *const dist_getc[] =
@@ -142,9 +150,7 @@
     PRIMITIVE(equal), PRIMITIVE(big), PRIMITIVE(eqbig),
     PRIMITIVE(startworld), PRIMITIVE(forkworld), PRIMITIVE(joinworld),
     PRIMITIVE(putc),
-#ifndef ALLOC_TRACE
     PRIMITIVE(errc),
-#endif
     PRIMITIVE(getc),
     PRIMITIVE_END
 };
@@ -153,12 +159,12 @@
 static int place_source(const char *path, const char **sip_p)
 {
     cell *cur = ram;
-    FILE *file = fopen(path, "r");
+    int file = my_open(path);
     char *sip = ram[0].m;
     size_t brac = 0;
     while (1) {
 	char *curp = cur->m;
-	size_t i, byte = fread(curp, 1, sizeof(cell), file);
+	size_t i, byte = my_read(file, curp, sizeof(cell));
 	for (i = 0; i < byte; i++)
 	    switch (curp[i]) {
 	    case '(':
@@ -176,7 +182,7 @@
 	}
 	cur++;
     }
-    fclose(file);
+    my_close(file);
     if (0 < brac)
 	return 2;
     cell_allocator_init(cur + 1, &ram[RAM_SIZE], NULL);
@@ -189,13 +195,20 @@
     *sip_p = sip;
     return 0;
 }
+static void print_error(const char *msg)
+{
+    size_t msg_len;
+    for (msg_len = 0; '\0' != msg[msg_len]; msg_len++);
+    my_write(my_stderr, msg, msg_len);
+}
 int main(int argc, char **argv)
 {
     //const char romsrc[] = ")";
     const char *ramsrc;
     int rst;
     if (2 != argc) {
-	fprintf(stderr, "%s [source file]\n", argv[0]);
+	print_error(argv[0]);
+	print_error(" [source file]\n");
 	return 1;
     }
     rst = place_source(argv[1], &ramsrc);
